[KAFKA-8820] [WIP] Ongoing work for changing the ReassignPartitionsCommand

steverod opened the pull request:
A work-in-progress collection of changes to implement KIP-455 for the ReassignPartitionsCommand.

These changes are intended to remove the dependence of the ReassignPartitionsCommand on the ZK client and utilize the new AdminClient interfaces for doing partition reassignments. Eventually the command will support incremental changes to ongoing reassignments, including cancellation.

This is being done in the background and will be updated irregularly (hence the WIP designation).

The changes are significant, but the commits are ordered to be relatively small and easy to understand and build on each other; it is strongly recommended that the review be done a commit at a time rather than all at once.

The first several commits are "infrastructure" work that gets things ready. Then a service abstraction which can be used with either ZK or Admin clients is provided and implemented, so that the command can (eventually) run entirely with just the Admin client. Unit tests are provided for the Service client (XXX: not all tests are implemented yet but those that exist should be fine.)

steverod committed:
Use getSortedBrokerList in removeThrottle instead of getAllBrokersInCluster.

We were doing getAllBrokersInCluster and then getting ids, which
internally just does getSortedBrokerList. So skip the extra work.

steverod committed:
Create ReassignCommandServiceTraits and provide a ZK-based client.

Abstract the functionality needed by the reassign partition command to be (eventually)
usable by either ZK or the Admin client, provide an initial implementation using the ZK client,
and switch the existing tests to work with the ZK-based Service Client.
(This causes problems for tests which do both ZK- and Admin-client based operations,
to be fixed in later commits.)

Add extra entry points to the ReassignPartitionCommand entrypoints.
Handle injection of mock AdminZkClients.
(Some of the test cases use heavily mocked adminZkClients, so adapt for that.)

steverod committed:
Make the ReassignPartitionsCommand have a private constructor.

This is (will be) useful for reducing the number of entry points as we move
away from always requiring the zkClient.

steverod committed:
Implement the AdminClient version of the ReassignCommandServiceClient.

Fill out the basic functionality needed for the ReassignPartitionsCommand when provided
by the AdminClient.

steverod committed:
Test fix: Some tests were sharing a common Properties object for brokers.

This meanth that when one set of "broker properties" was changed, it was changed for all brokers.
Ensure that all brokers have their own unique properties.

steverod committed:
Switch to full use of the service client.

Handle either ZK or Admin clients being provided by having overloaded "constructors"
(ReassignCommand apply methods) to create an appropriate service client. This change
means that you can either use the ZK client OR the Admin client but not both.

With the ServiceClient, only one of the ZK client or AdminClient is needed to access reassign
functionality. (Admin is preferred but ZK can continue to be used as in the past.)
Now the Service client holds all the functionality, and allow the ReassignPartitionsCommand
object to be created with either, one at a time.

steverod committed:
Implement unit tests for the service client, using either the ZK or Admin
versions. Tests are only provided for routines which are not trivial function
mappings (so the ZK test is somewhat abbreviated).

steverod committed:
Add a package-private ReassignPartitionsCommand constructor for testing.

This one takes a ServiceClient, which normally no one should care about.

steverod committed:
Modify the ReassignPartitionsCommandTest to work with mock Service client/Admin-based service client.

steverod committed:
Convert ReassignPartitionsClusterTest to primarily work using the AdminClient.

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -27,16 +28,226 @@ import kafka.utils._
+trait ReassignCommandService extends AutoCloseable {
+  def getBrokerIdsInCluster: Seq[Int]
+  def getBrokerMetadatas(rackAwareMode: RackAwareMode, brokerList : Option[Seq[Int]]) : Seq[BrokerMetadata]
+  def UpdateBrokerConfigs(broker : Int, configs : Map[String, String]) : Boolean
```
nit: capital letter in function name

steverod replied:
Fixed.

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -27,16 +28,226 @@ import kafka.utils._
+  def UpdateTopicConfigs(topic : String, configs: Map[String, String]) : Boolean
+  def getPartitionsForTopics(topics : immutable.Set[String]) : Map[String, Seq[Int]]
+  def getReplicaLogDirsForTopics(topics : Set[TopicPartitionReplica]): Map[TopicPartitionReplica, ReplicaLogDirInfo]
+  def alterPartitionAssignment(topics: Map[TopicPartition, Seq[Int]], timeoutMs : Long) : Unit // XXX: Maybe not unit?
```
I think it'd be good user experience to return a map of results. The script can then log any errors for specific partitions.
Note the KIP says we could have the following errors in the response:
```
Possible partition-level errors:
INVALID_REPLICA_ASSIGNMENT - when the replica assignment is not valid (brokers are not part of the cluster, negative replica ids, empty assignment)
NO_REASSIGNMENT_IN_PROGRESS - when a cancellation was called on a topic/partition which was not in the middle of a reassignment
UNKNOWN_TOPIC_OR_PARTITION - when a topic/partition doesn't exist or a topic deletion is in progress
```

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -27,16 +28,226 @@ import kafka.utils._
+  def getOngoingReassignments: Map[TopicPartition, Seq[Int]]
+}
+
+case class AdminClientReassignCommandService  (adminClient : Admin) extends ReassignCommandService {
```
nit: space between AdminClientReassignCommandService and ()

steverod replied:
fixed.

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -27,16 +28,226 @@ import kafka.utils._
+
+    // Somewhat convoluted, because brokerList can either be None or an empty List, and in either
+    // case it's equivalent to "all brokers"
+    val relevantBrokers = clusterDescription.collect { case n : Node if brokerList.getOrElse(List()).contains(n.id) => n} match {
```
Could we refactor this? In what case does `clusterDescription.collect` not return a `n: Node`? I don't think this should return `Nil` - we prefer `None` in Scala

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -27,16 +28,226 @@ import kafka.utils._
+        case "" => AlterConfigOp.OpType.DELETE
+        case _ => AlterConfigOp.OpType.SET
+    })
+  }.asJavaCollection
```
nit: indent

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -27,16 +28,226 @@ import kafka.utils._
+    ).asJava)
+
+    alterResult.all().get()
+    // XXX How to determine success??
```
as far as I know, this will throw an exception if any one of the configs are invalid

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -27,16 +28,226 @@ import kafka.utils._
+    adminClient.alterPartitionReassignments(partitionReassignment.map{
+      entry  => (entry._1,
+        Some(new NewPartitionReassignment(seqAsJavaList(entry._2.map(brokerId => brokerId: java.lang.Integer)))).asJava)
+    }.asJava).all().get()
```
This call will also fail if we have an error for any one partition. Given the fact that the broker will execute the successful partitions - I think the reassign-partitions tool should handle that and suggest the user re-runs only the failed partitions

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -27,16 +28,226 @@ import kafka.utils._
+    for (topicData <- topicDescriptions) {
+      val topic = topicData._1
+      for (partition <- topicData._2.partitions.asScala) {
+        replicaList(new TopicPartition(topic, partition.partition())) = partition.replicas.asScala.map(_.id())
```
nit: I think convention is to use put or `+= (key -> val)`

stanislavkozlovski commented on clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java:
```diff
@@ -101,12 +101,6 @@ public void addTopic(boolean internal,
                 throw new IllegalArgumentException("Unknown brokers in isr list");
             }
-
-            if (replicas == null) {
```
Why is this?

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,9 +16,10 @@
+import java.util.{Collections, Properties}
+import java.util.concurrent.{ExecutionException, TimeUnit}
 
+import kafka.admin.ConfigCommand.ConfigEntity
```
nit: unused config

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,9 +16,10 @@
-import java.util.Properties
-import java.util.concurrent.ExecutionException
+import java.util.{Collections, Properties}
+import java.util.concurrent.{ExecutionException, TimeUnit}
```
nit: both additions are unused

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -48,26 +259,46 @@ object ReassignPartitionsCommand extends Logging {
+  def apply(adminClient:Admin,
+            proposedPartitionAssignment: Map[TopicPartition, Seq[Int]],
+            proposedReplicaAssignment: Map[TopicPartitionReplica, String]) : ReassignPartitionsCommand =
+    new ReassignPartitionsCommand(new AdminClientReassignCommandService(adminClient), proposedPartitionAssignment, proposedReplicaAssignment)
```
nit: `new` isn't needed in the `AdminClientReassignCommandService` construction

stanislavkozlovski commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsCommandTest.scala:
```diff
@@ -27,54 +27,106 @@ import kafka.utils.CoreUtils._
+  // Base Test Service Client with minimal functionality, just enough for our tests to
+  // override
+  class TestServiceClientBase extends ReassignCommandService {
+    override def getBrokerIdsInCluster: Seq[Int] = throw new UnimplementedException()
```
Should we use our own exception? I don't think we should depend on a ZK library exception

stanislavkozlovski commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsServiceClientTest.scala:
```diff
@@ -0,0 +1,394 @@
+import java.util.{Collections, Properties}
+
+import kafka.admin.{AdminClientReassignCommandService, AdminOperationException, BrokerMetadata, RackAwareMode, ZkClientReassignCommandService}
+import kafka.log.LogConfig.{FollowerReplicationThrottledReplicasProp, LeaderReplicationThrottledReplicasProp}
```
nit: unused

stanislavkozlovski commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsServiceClientTest.scala:
```diff
@@ -0,0 +1,394 @@
+import org.apache.kafka.clients.admin.{Admin, MockAdminClient}
+import org.apache.kafka.common.errors.UnknownTopicOrPartitionException
+import org.apache.kafka.common.{Node, TopicPartition, TopicPartitionInfo}
+import org.easymock.EasyMock.{replay, _}
```
There is an initiative to use `Mockito` as Kafka's mocking library, so we shouldn't introduce new tests with EasyMock

stanislavkozlovski commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsServiceClientTest.scala:
```diff
@@ -0,0 +1,394 @@
+
+    val serviceTest3 = new ZkClientReassignCommandService(mockZkClient, Some(mockAdminZkClient))
+
+    serviceTest3.UpdateTopicConfigs("TOPIC0", configUpdatesMap)
```
are we missing a `replay()` here?

stanislavkozlovski commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsServiceClientTest.scala:
```diff
@@ -0,0 +1,394 @@
+
+    val serviceTest3 = new ZkClientReassignCommandService(mockZkClient, Some(mockAdminZkClient))
+
+    serviceTest3.UpdateBrokerConfigs(testBrokerId, configUpdatesMap)
```
ditto with the replay

stanislavkozlovski commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsServiceClientTest.scala:
```diff
@@ -0,0 +1,394 @@
+/*
```
I like this test class! Thanks for adding it

One suggestion I have is to prefer smaller test methods that have descriptive names with regards on what they test. This helps narrow down failures as fast as possible (ideally, the method name should tell you). This was highly recommended in [Growing Object-Oriented Software, Guided by Tests (Beck Signature)](https://www.amazon.co.uk/Growing-Object-Oriented-Software-Guided-Signature/dp/0321503627) which I found was a very good book

To give a concrete example: I'd change the `testGetReplicaAssignmentForTopics` method to be 3 separate tests:
`testGetReplicaAssignmentForTopicsShouldFetchPartitionsForAllKnownTopics`
`testGetReplicaAssignmentForTopicsShouldNotFetchAllAvailableTopics`
`testGetReplicaAssignmentForTopicsShouldThrowExceptionOnUnknownTopic`

This results in long test method names but that's a feature IMO - people skimming over the test quickly realize what's tested vs what's not - and test failures suddenly become very clear (in the IDE or CI), since the test method clearly communicates what's failed

stanislavkozlovski commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsServiceClientTest.scala:
```diff
@@ -0,0 +1,394 @@
+    // Note the non-sequential, non-zero-based IDs.
+    val mockAdmin = new MockAdminClient(brokerList.asJava, brokerList(2))
+
+    mockAdmin.addTopic(false, "test1", test1Partitions, Collections.emptyMap())
```
Could we move the `test1` and etc topic names to variables? That way we better know what topics the admin client from `setUpTestCluster()`  supports, rather than flying strings

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -48,26 +259,46 @@ object ReassignPartitionsCommand extends Logging {
+            proposedReplicaAssignment: Map[TopicPartitionReplica, String]) : ReassignPartitionsCommand =
+    new ReassignPartitionsCommand(serviceClient, proposedPartitionAssignment, proposedReplicaAssignment)
+
   def main(args: Array[String]): Unit = {
```
We need to add the cancel functionality to cancel a reassignment

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -196,26 +431,33 @@ object ReassignPartitionsCommand extends Logging {
 
     // If there is an existing rebalance running, attempt to change its throttle
-    if (zkClient.reassignPartitionsInProgress()) {
+    if (serviceClient.reassignInProgress) {
```
I'm not sure we want to limit this when we have a reassignment in progress. The goal of KIP-455 is to allow incremental reassignments

stanislavkozlovski reviewed (commented):
I gave this a brief overview. I think we should separate the AdminClient/ZK tests in `ReassignPartitionsCommandTest` more explicitly (not by code comments). My suggestion would be by test method names.

I think we want to add some deprecated mention in the `--zookeeper` functionality.
We need to support a `--create-cancellation-plan` and `--list`

I see our `ReassignPartitionsCommandTest` sometimes exercises mocks and sometimes tests the actual thing. My personal preference would be to separate both test types into different classes such that we can more easily tell how much actual tests we cover by just reading the method names. As it stands, we need to go over each test method's implementation to understand how much is mocked versus how much is not.
Interested to see what others think
I think this PR should also add a system test for KIP-455, as we weren't able to otherwise. This would also help validate the tool works correctly. I suggest reworking the `reassign_partitions_test.py` to work with both ZK and the new API

steverod committed:
API cleanup: Remove unneeded ZK test accesses, unneeded parameters.

Add APIs for ReassignPartitionsCommand object that take AdminClients (preferred)
and use those is tests.

steverod committed:
AdminClient: getBrokerMetadatas requires Spotbugs exemption for Scala 2.12 compiler.

steverod committed:
ServiceClientTest: Split out multi-case tests.

steverod committed:
Close down some methods to package/Kafka-private for testing.

steverod committed:
Remove unnecessary method.

steverod committed:
Convert a method to using the Service Client (missed originally).

steverod committed:
Clarify a comment that's now using the service client.

steverod committed:
Remove unnecessary Option parameters from verifyAssignments.

The AdminClientOpt was needed initially but is now superseded by the ServiceClient.

steverod committed:
Remove AdminClientOpt from execute assignment -- no longer needed.

Migrate some extra tests away from the ZK client form of executeAssignment.

steverod committed:
Make ReassignPartitionsCommand main() use AdminClient preferentially.

Fall back to ZK if there is no bootstrap-server specified.

steverod committed:
Adjust method visibility again.

adminClient-taking methods are public, but the ServiceClient is / ought to be reassign-command internal.
ZK should not be exposed in general.

steverod committed:
Mark a number of ZK operations as deprecated.

steverod committed:
Remove some unneeded imports.

steverod committed:
Remove unused local variable from ReassignPartitionsCommandTest.

steverod committed:
Review feedback: Rename Update*Configs to update*configs (proper Java casing).

steverod committed:
Review feedback: whitespace, more idiomatic way of adding keys to a map.

steverod committed:
Review feedback: Replace the ZK UnimplementedException with the Java-standard UnsupportedOperationException.

steverod committed:
Review feedback: Change test topic names from hardcoded strings to test-class constants.

cmccabe commented:
The reassign-partitions.sh command was upgraded to use the `--bootstrap-server` flag in #8244
